Attach comments to AST nodes.

laurentlb opened the pull request:
This feature is off by default. More testing will be needed before
exposing it to the ParseFile function.

Logic is copied from Buildifier (https://github.com/bazelbuild/buildtools/tree/master/build).

bug #63

alandonovan commented on syntax/parse.go:
```diff
@@ -16,6 +16,9 @@ import "log"
 const debug = false
 
+// Enable this flag to attach comments to the AST.
+const keepComments = false
```
Naturally this should be a mode flag to the parser, not a compile-time constant.  Add:

```
  type Mode uint
  const (
     RetainComments Mode = 1<<iota
  )
  func Parse(..., mode Mode)
  func ParseExpr(..., mode Mode)
```

This will be a breaking API change, but so be it.

laurentlb replied:
Done

alandonovan commented on syntax/parse.go:
```diff
@@ -66,16 +72,25 @@ func ParseExpr(filename string, src interface{}) (expr Expr, err error) {
 	p.tok = p.in.nextToken(&p.tokval)
+	// save comments
+	for p.tok == LINE_COMMENT || p.tok == SUFFIX_COMMENT {
+		if keepComments {
```
Pass the comments flag down into the scanner, so that p.in.nextToken never returns comment tokens unless that mode is enabled.

alandonovan commented on syntax/parse.go:
```diff
@@ -938,3 +953,99 @@ func terminatesExprList(tok Token) bool {
+
+// flattenAst returns the list of AST nodes, both in prefix order and in postfix
+// order.
+func flattenAst(f *File) (pre []Node, post []Node) {
```
flattenAST

(Go uses the same case for all letters of an initialism.)

laurentlb replied:
Done

alandonovan commented on syntax/parse.go:
```diff
@@ -938,3 +953,99 @@ func terminatesExprList(tok Token) bool {
+
+// flattenAst returns the list of AST nodes, both in prefix order and in postfix
+// order.
+func flattenAst(f *File) (pre []Node, post []Node) {
```
Factor: pre, post []Node

laurentlb replied:
Done

alandonovan commented on syntax/parse.go:
```diff
@@ -938,3 +953,99 @@ func terminatesExprList(tok Token) bool {
+// flattenAst returns the list of AST nodes, both in prefix order and in postfix
+// order.
+func flattenAst(f *File) (pre []Node, post []Node) {
+	pre = []Node{}
```
These two assignments are unnecessary.  The zero value of a slice is a valid empty slice.

laurentlb replied:
Done

alandonovan commented on syntax/parse.go:
```diff
@@ -938,3 +953,99 @@ func terminatesExprList(tok Token) bool {
+		}
+	}
+
+	// // Assign line comments to syntax immediately following.
```
doubled //

laurentlb replied:
Done

alandonovan commented on syntax/scan.go:
```diff
@@ -196,6 +200,13 @@ func (p Position) String() string {
 	return fmt.Sprintf("%s:%d:%d", p.Filename(), p.Line, p.Col)
 }
 
+func (p Position) IsBefore(p2 Position) bool {
```
This method needn't be exported: isBefore

Call p2 "q".

laurentlb replied:
Done

alandonovan commented on syntax/scan.go:
```diff
@@ -462,11 +473,18 @@ start:
 
 	// comment
 	if c == '#' {
+		sc.startToken(val)
```
The scanner should know whether comments are required, and skip the new logic if so.

alandonovan commented on syntax/scan.go:
```diff
@@ -462,11 +473,18 @@ start:
 			sc.readRune()
 			c = sc.peekRune()
 		}
+		sc.endToken(val)
```
Skip this unless comments are requested.  It allocates a string for the comment.

alandonovan commented on syntax/syntax.go:
```diff
@@ -9,6 +9,33 @@ package syntax
 	// Span returns the start and end position of the expression.
 	Span() (start, end Position)
+
+	Comment() *Comments
```
// Comments returns the comments associated with this node.
// It returns nil if RetainComments was not specified during parsing.

(see below)

laurentlb replied:
Done

alandonovan commented on syntax/syntax.go:
```diff
@@ -9,6 +9,33 @@ package syntax
+// A Comment represents a single # comment.
+type Comment struct {
+	Start  Position
+	Token  string // without trailing newline
```
Text string

alandonovan commented on syntax/syntax.go:
```diff
@@ -9,6 +9,33 @@ package syntax
+// Comments collects the comments associated with an expression.
+type Comments struct {
+	Before []Comment // whole-line comments before this expression
+	Suffix []Comment // end-of-line comments after this expression
```
Can there be more than one of these? I assume not.

Add "(up to 1)" to the doc comment.

alandonovan replied:
Not yet done.

alandonovan commented on syntax/syntax.go:
```diff
@@ -25,6 +52,7 @@ func End(n Node) Position {
 
 // A File represents a Skylark file.
 type File struct {
+	Comments
```
Please make each of these a pointer.  In the performance critical case (BUILD file loading), these Comment fields add 9 words (3 slices x 3 words) to every AST node, which more than doubles the size of the AST.  Making these fields pointers would add only 1 word to each node.  This does require the parser to make additional allocations when comments are requested, but applications should pay only for what they use (and reformatting tools typically load only a small handful of BUILD files in their lifetime, not tens of thousands).

alandonovan replied:
Of course, this assumes that Skylark comments will never become significant to the semantics, as they have (regrettably) in Go.

laurentlb replied:
Right, we don't intend to make comments significant.
(they might affect linters and other tools though, but not the interpreter)

alandonovan commented on syntax/syntax.go:
```diff
@@ -9,6 +9,33 @@ package syntax
+type Comment struct {
+	Start  Position
+	Token  string // without trailing newline
+	Suffix bool   // an end of line (not whole line) comment
```
"end-of-line"

laurentlb replied:
Done

alandonovan commented on syntax/syntax.go:
```diff
@@ -9,6 +9,33 @@ package syntax
+type Comment struct {
+	Start  Position
+	Token  string // without trailing newline
+	Suffix bool   // an end of line (not whole line) comment
```
Suffix comments are handled differently from line comments in all places, so rather than put an extra bit (actually a word) in every comment, just keep them separate.  Make the parser have, in place of one field p.comments, two fields lineComments (a slice) and suffixComment (an optional one).

laurentlb replied:
As I try to converge this AST with Buildifier's AST, I'd prefer to keep the type like this for now. I'll check what Buildifier exactly does before changing the structure.

laurentlb commented:
Good point! I've removed the new tokens and simplified the code.

PTAL

alandonovan commented on syntax/parse.go:
```diff
@@ -16,15 +16,21 @@ import "log"
 // Enable this flag to print the token stream and log.Fatal on the first error.
 const debug = false
 
+type Mode uint
```
// A Mode value is a set of flags (or 0) that controls optional parser functionality.

laurentlb replied:
Done

alandonovan commented on syntax/parse.go:
```diff
@@ -16,15 +16,21 @@ import "log"
+type Mode uint
+
+const (
+	RetainComments Mode = 1 << iota
```
RetainComments Mode = 1 << iota // retain comments in AST; see Node.Comments

laurentlb replied:
Done

alandonovan commented on syntax/parse.go:
```diff
@@ -36,13 +42,16 @@ func Parse(filename string, src interface{}) (f *File, err error) {
-func ParseExpr(filename string, src interface{}) (expr Expr, err error) {
-	in, err := newScanner(filename, src)
+func ParseExpr(filename string, src interface{}, mode Mode) (expr Expr, err error) {
+	in, err := newScanner(filename, src, mode&RetainComments != 0)
```
Doesn't this function need a call to p.assignComments? How to comments get attached?

laurentlb replied:
Done

alandonovan commented on syntax/parse.go:
```diff
@@ -938,3 +947,99 @@ func terminatesExprList(tok Token) bool {
+
+// flattenAST returns the list of AST nodes, both in prefix order and in postfix
+// order.
+func flattenAST(f *File) (pre, post []Node) {
```
s/f *File/n Node/

laurentlb replied:
Done

laurentlb replied:
Done

alandonovan commented on syntax/parse.go:
```diff
@@ -938,3 +947,99 @@ func terminatesExprList(tok Token) bool {
 }
+
+// Comment assignment.
+// We build two lists of all subexpressions, preorder and postorder.
```
The term "expression" in this paragraph (from buildifier) should be replaced with node.

laurentlb replied:
Done

alandonovan commented on syntax/parse.go:
```diff
@@ -938,3 +947,99 @@ func terminatesExprList(tok Token) bool {
+}
+
+// assignComments attaches comments to nearby syntax.
+func (p *parser) assignComments(f *File) {
```
s/f *File/n Node/

laurentlb replied:
Done

alandonovan commented on syntax/parse.go:
```diff
@@ -938,3 +947,99 @@ func terminatesExprList(tok Token) bool {
+
+	// Split into whole-line comments and suffix comments.
+	var line, suffix []Comment
+	for _, com := range p.in.comments {
```
Why not have the scanner divide the comments into the two categories?
Then you can do away with the .Suffix field.

laurentlb replied:
We need to keep the .Suffix field in the AST. This is a useful information, for example if we want to print it back.

laurentlb replied:
Well... the information seems duplicated (there's a Suffix Comment slice and the suffix boolean field). I have to investigate why Buildifier does it like this.

alandonovan commented on syntax/parse.go:
```diff
@@ -938,3 +947,99 @@ func terminatesExprList(tok Token) bool {
+	// Assign line comments to syntax immediately following.
+	for _, x := range pre {
+		start, _ := x.Span()
+		x.AllocComments()
```
Don't call AllocComments until you need to!

laurentlb replied:
Done

alandonovan commented on syntax/parse.go:
```diff
@@ -938,3 +947,99 @@ func terminatesExprList(tok Token) bool {
+	}
+
+	// Remaining line comments go at end of file.
+	f.AllocComments()
```
Don't call AllocComments unless line is non-empty.

laurentlb replied:
Done

alandonovan commented on syntax/parse.go:
```diff
@@ -938,3 +947,99 @@ func terminatesExprList(tok Token) bool {
+		}
+
+		_, end := x.Span()
+		xcom := x.Comments()
```
Call AllocComments before Comments, and don't call either until you first need to.

laurentlb replied:
Done

alandonovan commented on syntax/parse.go:
```diff
@@ -938,3 +947,99 @@ func terminatesExprList(tok Token) bool {
+	}
+
+	// We assigned suffix comments in reverse.
+	// If multiple suffix comments were appended to the same
```
How can there be multiple suffix comments for a single node?

laurentlb replied:
I don't know, this needs investigation.

alandonovan commented on syntax/parse.go:
```diff
@@ -938,3 +947,99 @@ func terminatesExprList(tok Token) bool {
+	}
+
+	// Remaining suffix comments go at beginning of file.
+	f.Comments().Before = append(f.Comments().Before, suffix...)
```
Call AllocComments first.

laurentlb replied:
Done

alandonovan commented on syntax/parse_test.go:
```diff
@@ -257,6 +258,7 @@ func stripPos(err error) string {
 func treeString(n syntax.Node) string {
 	var buf bytes.Buffer
 	writeTree(&buf, reflect.ValueOf(n))
+	// Skip comments fields
```
This comment doesn't belong here.  Delete it?

laurentlb replied:
Done

alandonovan commented on syntax/scan.go:
```diff
@@ -196,29 +196,39 @@ func (p Position) String() string {
+	indentstk    []int     // stack of indentation levels
+	dents        int       // number of saved INDENT (>0) or OUTDENT (<0) tokens to return
+	lineStart    bool      // after NEWLINE; convert spaces to indentation tokens
+	keepComments bool      // if false, do not return comments tokens
```
This comment is stale. Also, document a boolean using a verb phrase:

// accumulate comments in slice

laurentlb replied:
Done

alandonovan commented on syntax/scan.go:
```diff
@@ -196,29 +196,39 @@ func (p Position) String() string {
+	dents        int       // number of saved INDENT (>0) or OUTDENT (<0) tokens to return
+	lineStart    bool      // after NEWLINE; convert spaces to indentation tokens
+	keepComments bool      // if false, do not return comments tokens
+	comments     []Comment // list of comments seen in the file (if keepComments was true)
```
Divide this into two:

lineComments, suffixComments []Comment

alandonovan commented on syntax/scan.go:
```diff
@@ -442,6 +452,7 @@ start:
 	if sc.dents != 0 {
 		sc.startToken(val)
 		sc.endToken(val)
+		blank = false
```
The change is no longer necessary.

laurentlb replied:
Done

alandonovan commented on syntax/scan.go:
```diff
@@ -475,12 +494,14 @@ start:
 		sc.startToken(val)
 		sc.readRune()
 		val.raw = "\n"
+		blank = false
```
Delete

alandonovan commented on syntax/scan.go:
```diff
@@ -475,12 +494,14 @@ start:
 		if blank || sc.depth > 0 {
 			// Ignore blank lines, or newlines within expressions (common case).
 			sc.readRune()
+			blank = false
```
Delete

laurentlb replied:
Done

alandonovan commented on syntax/syntax.go:
```diff
@@ -9,6 +9,47 @@ package syntax
 	Span() (start, end Position)
+
+	// Comments returns the comments associated with this node.
+	// It returns nil if RetainComments was not specified during parsing.
```
"during parsing, or if AllocComments was not called."

laurentlb replied:
Done

alandonovan commented on syntax/syntax.go:
```diff
@@ -9,6 +9,47 @@ package syntax
+type Comment struct {
+	Start  Position
+	Text   string // without trailing newline
+	Suffix bool   // an end-of-line (not whole line) comment
```
delete

alandonovan commented on syntax/syntax.go:
```diff
@@ -9,6 +9,47 @@ package syntax
+	After []Comment
+}
+
+// A CommentsRef is a possibly-nil reference to a set of comments.
```
type commentsRef (lowercase)

laurentlb replied:
This breaks the tests:

--- FAIL: TestExprParseTrees (0.00s)
panic: reflect.Value.Interface: cannot return value obtained from unexported field or method [recovered]
	panic: reflect.Value.Interface: cannot return value obtained from unexported field or method

alandonovan replied:
The loop over struct fields in treeString will need to skip the commentRef field.

alandonovan commented on syntax/parse.go:
```diff
@@ -16,15 +16,21 @@ import "log"
 // Enable this flag to print the token stream and log.Fatal on the first error.
 const debug = false
 
+type Mode uint // A Mode value is a set of flags (or 0) that controls optional parser functionality.
```
This comment should precede the declaration.

alandonovan commented on syntax/parse.go:
```diff
@@ -938,3 +950,103 @@ func terminatesExprList(tok Token) bool {
+		}
+
+		_, end := x.Span()
+		for len(suffix) > 0 && end.isBefore(suffix[len(suffix)-1].Start) {
```
It would be simpler and more efficient to find the least index for which the isBefore condition holds, then make a single call to AllocComments and append. You also wouldn't need to reverse the slice later.

alandonovan replied:
Still not done.

laurentlb replied:
Well, I've removed the reverse. If we assume there's at most one suffix comment, loop should be done 0 or 1 time. I don't think we can make it more efficient.

alandonovan replied:
We'll, we can avoid unnecessarily re-checking the loop condition: replace "for" with "if".

laurentlb commented:
PTAL (yes, GitHub is not the best tool for code reviews).
I've removed `Suffix bool`, renamed `commentsRef`, removed the list reversing in `assignComments`.

alandonovan commented on syntax/parse.go:
```diff
@@ -938,3 +951,81 @@ func terminatesExprList(tok Token) bool {
+
+// assignComments attaches comments to nearby syntax.
+func (p *parser) assignComments(n Node) {
+	pre, post := flattenAST(n)
```
Optimization: `if len(p.in.lineComments) + len(p.in.suffixComments) == 0`, you can skip the rest of this function, in particular the flattenAST step.  Once you've done that, it's no longer necessary to guard the two calls to assignComments with `if mode&retainComments != 0`.

alandonovan commented on syntax/scan.go:
```diff
@@ -196,29 +196,40 @@ func (p Position) String() string {
+	indentstk      []int     // stack of indentation levels
+	dents          int       // number of saved INDENT (>0) or OUTDENT (<0) tokens to return
+	lineStart      bool      // after NEWLINE; convert spaces to indentation tokens
+	keepComments   bool      // /accumulate comments in slice
```
rogue '/'

alandonovan commented on syntax/scan.go:
```diff
@@ -196,29 +196,40 @@ func (p Position) String() string {
+	dents          int       // number of saved INDENT (>0) or OUTDENT (<0) tokens to return
+	lineStart      bool      // after NEWLINE; convert spaces to indentation tokens
+	keepComments   bool      // /accumulate comments in slice
+	lineComments   []Comment // list of full line comments (if keepComments was true)
```
strike " was true", x 2

laurentlb commented:
PTAL. Addressed all comments (I think), except the one where I replied.

alandonovan commented on syntax/parse.go:
```diff
@@ -938,3 +946,86 @@ func terminatesExprList(tok Token) bool {
+		}
+	}
+
+	// Remaining suffix comments go at beginning of file.
```
Under what circumstances can there be remaining suffix comments?

laurentlb commented:
Done.

I will need to do more testing and iterate on this (in a separate PR).

For the record, there are a few differences between this PR and Buildifier behavior.

1. Buildifier supports multiple suffix comments, e.g.

```
greeting = ("hello" +  # c20
            # c19
            "world") 
```

`c19` and `c20` are both suffix comments attached to the same node. In this PR, `c19` is a line comment instead.

2. Buildifier has support for CommentBlock represented as a separate node (like a statement).

alandonovan commented:
1. Interesting point about multiple suffix comments.  I wonder if they're necessary?  What do we lose by representing c19 as a Before comment of "world".

2. I noticed CommentBlock; but again, is it necessary?  I would expect that a bunch of Before or After comments with discontiguous line numbers would be sufficient to represent the structure in all its detail.

Russ might remember the design rationale here.

laurentlb commented:
Thanks for the thorough review, it was very useful!

Multiple suffix comments don't seem useful to me. The idea of CommentBlock is that they are not attached to anything. So if you move/remove the function after a block, the comment stays. This has been useful for many cleanup I've done in the past.

alandonovan commented:
I failed to notice that the new code doesn't have any tests.  Could you write a test that the parser produces ASTs with comment fields populated in the expected way, and feed it some tricky inputs: comments interspersed with function call arguments, blocks of comments between statements, and so on?  Thanks.

laurentlb commented:
Sure.
